internal/archive/extract.go

```diff
@@ -21,22 +26,51 @@ const (
 	mimeTypeGzip = mimeType("application/gzip")
 )
 
-type Extractor func(url string) (*Extraction, error)
+var defaultBackoff = wait.Backoff{
+	Duration: time.Second,
+	Factor:   2,
+	Steps:    10,
+	Jitter:   0.1,
+	Cap:      30 * time.Second,
+}
+
+type fileDownloader interface {
+	Get(string) (*http.Response, error)
+}
+
+type Extractor func(logr.Logger, context.Context, string, time.Duration) (*Extraction, error)
 
 type Extraction struct {
 	RootDir     string
 	Archive     string
 	ContentsDir string
 }
 
-func FetchAndExtract(url string) (*Extraction, error) {
+func FetchAndExtract(log logr.Logger, ctx context.Context, url string, timeout time.Duration) (*Extraction, error) {
+	if timeout > 0 {
+		var cancel context.CancelFunc
+		ctx, cancel = context.WithTimeout(ctx, timeout)
+		defer cancel()
+	}
```

Comment:
I don't think this matters now but it might be a good idea to start the timeout right before the call in case someone decided to _insert_ lengthy operations between here and there.